src/main/java/com/google/api/generator/engine/ast/JavaDocComment.java

```diff
@@ -0,0 +1,107 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package com.google.api.generator.engine.ast;
+
+import com.google.auto.value.AutoValue;
+import com.google.common.collect.ImmutableList;
+import java.util.ArrayList;
+import java.util.List;
+import javax.annotation.Nullable;
+
+@AutoValue
+public abstract class JavaDocComment implements Comment {
+
+  @Nullable
+  public abstract String deprecated();
+
+  @Nullable
+  public abstract String throwsText();
+
+  public abstract ImmutableList<String> comments();
+
+  public static Builder builder() {
+    return new AutoValue_JavaDocComment.Builder();
+  }
+
+  @AutoValue.Builder
+  public abstract static class Builder {
+    protected abstract ImmutableList.Builder<String> commentsBuilder();
+
+    public abstract Builder setDeprecated(String deprecatedText);
+
+    public abstract Builder setThrowsText(String throwsText);
+
+    public Builder addComment(String comment) {
+      commentsBuilder().add(String.format("%1$s" + comment + "%2$s", "* ", "\n"));
+      return this;
+    }
+
+    public Builder addParam(String name, String description) {
+      String parameter =
+          String.format("%1$s" + name + "%2$s" + description + "%3$s", "* @param ", " ", "\n");
+      commentsBuilder().add(parameter);
+      return this;
+    }
+
+    public Builder addSampleCode(String sampleCode) {
+      commentsBuilder().add("* Sample code:\n* <pre><code>\n");
+      String[] sampleLines = sampleCode.split("\\r?\\n");
+      for (int i = 0; i < sampleLines.length; i++) {
+        sampleLines[i] = "* " + sampleLines[i];
+      }
+      commentsBuilder()
+          .add(String.format(String.join("\n", sampleLines) + "%s", "\n* </code></pre>\n"));
+      return this;
+    }
+
+    public Builder addHtmlP(String paragraph) {
+      commentsBuilder().add(String.format("%1$s" + paragraph + "%2$s", "* <p> ", "\n"));
+      return this;
+    }
+
+    public Builder addHtmlOl(List<String> oList) {
+      commentsBuilder().add("* <ol>\n");
+      for (int i = 0; i < oList.size(); i++) {
+        oList.set(i, String.format("%1$s" + oList.get(i) + "%2$s", "* <li>", "\n"));
+      }
+      commentsBuilder().add(String.format(String.join("", oList) + "%s", "* </ol>\n"));
+      return this;
+    }
+
+    public Builder addHtmlUl(List<String> uList) {
+      commentsBuilder().add("* <ul>\n");
+      for (int i = 0; i < uList.size(); i++) {
+        uList.set(i, String.format("%1$s" + uList.get(i) + "%2$s", "* <li>", "\n"));
+      }
+      commentsBuilder().add(String.format(String.join("", uList) + "%s", "* </ul>\n"));
+      return this;
+    }
+
+    public abstract JavaDocComment build();
+  }
+
+  @Override
+  public String comment() {
+    List<String> commentBody = new ArrayList<>(comments());
```

Comment:
Can we do this without instantiating a new list?

Reply:
yes, found a way to convert to stream and collect as a list.